pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,296 @@
+		if value == nil {
+			return true
+		}
+		v := value.(ring.KVEntry)
+		cluster := v.Cluster
+		replica := v.ReplicaEntry.(*ReplicaDesc)
+		c.setReplica(cluster, replica.GetReplica(), timestamp.Time(replica.GetTimestamp()))
+		return true
+	})
+}
+
+// Stop ends calls the trackers cancel function, which will end the loop for WatchPrefix.
+func (c *tracker) Stop() {
+	c.cancel()
+	<-c.done
+}
+
+// Wraps calls to client CAS with locking and error checking
+func (c *tracker) casWrapper(ctx context.Context, cluster, replica string, now time.Time) error {
+	// We have should a prefixed KVStore client, so we want to pass /cluster instead of just cluster.
+	// Prefixed client does prefix+key without separating the values.
+	err := c.client.CAS(ctx, fmt.Sprintf("/%s", cluster), func(in interface{}) (out interface{}, retry bool, err error) {
+		desc, ok := in.(*ReplicaDesc)
+		// There was either invalid or no data for the key, so we now accept samples
+		// from this replica.
+		if !ok || desc == nil {
+			desc = &ReplicaDesc{
+				Replica:   replica,
+				Timestamp: timestamp.FromTime(now),
+			}
+			return desc, false, nil
+		}
+
+		// As far as the timestamp in the KVStore is concerned none of the Distributors have
+		// received a sample from the stored replica recently, so we should overwrite the
+		// replica in the KVStore and accept this sample.
+		if now.Sub(timestamp.Time(desc.Timestamp)) > c.overwriteTimeout {
+			desc = &ReplicaDesc{
+				Replica:   replica,
+				Timestamp: timestamp.FromTime(now),
+			}
+			return desc, false, nil
+		}
+
+		// Safeguard against other calls to the KVStore having changed the accepted replica via a
+		// CAS to the KVStore before us. We should only return a replica description when the
+		// replicas don't match in the case above, where the overwrite timeout has been reached.
+		if desc.Replica != replica {
+			level.Info(util.Logger).Log("msg", "replicas did not match in CAS call")
+			return nil, false, fmt.Errorf("replicas did not match, rejecting sample: %s != %s", replica, desc.Replica)
+		}
+
+		// As far as the timestamp in the KVStore is concerned none of the Distributors have
+		// updated the timestamp for this cluster/replica in more than the write timeout.
+		// We should update the timestamp so that we don't potentially failover to accepting
+		// samples from another replica when we don't need to.
+		// This may only catch local CAS' updating the KVStore, not other Distributors.
+		if now.Sub(timestamp.Time(desc.Timestamp)) > c.writeTimeout {
+			desc = &ReplicaDesc{
+				Replica:   replica,
+				Timestamp: timestamp.FromTime(now),
+			}
+			return desc, false, nil
+		}
+		// TODO: should this just return nil for an error as well? It results in an error log when
+		// we return ErrDontOverwrite. Here the cluster/replica matched, but we just don't want to
+		// update what's in the KVStore so we return nil for the interface.
+		return nil, false, &ErrDontOverwrite{}
```

Comment:
Still don't like `ErrDontOverwrite`, I think updating CAS so a nil value means dont' write is more correct.  Errors are for errors, not success.